sycl/include/CL/sycl/builtins.hpp

```diff
@@ -1395,175 +1530,284 @@ select(T a, T b, T2 c) __NOEXC {
   return __sycl_std::__invoke_native_log<T>(x);
 }
 
 // genfloatf log2 (genfloatf x)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> log2(T x) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> log2(T x) __NOEXC {
   return __sycl_std::__invoke_native_log2<T>(x);
 }
 
 // genfloatf log10 (genfloatf x)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> log10(T x) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> log10(T x) __NOEXC {
   return __sycl_std::__invoke_native_log10<T>(x);
 }
 
 // genfloatf powr (genfloatf x, genfloatf y)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> powr(T x, T y) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> powr(T x,
+                                                              T y) __NOEXC {
   return __sycl_std::__invoke_native_powr<T>(x, y);
 }
 
 // genfloatf recip (genfloatf x)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> recip(T x) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> recip(T x) __NOEXC {
   return __sycl_std::__invoke_native_recip<T>(x);
 }
 
 // genfloatf rsqrt (genfloatf x)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> rsqrt(T x) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> rsqrt(T x) __NOEXC {
   return __sycl_std::__invoke_native_rsqrt<T>(x);
 }
 
 // genfloatf sin (genfloatf x)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> sin(T x) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> sin(T x) __NOEXC {
   return __sycl_std::__invoke_native_sin<T>(x);
 }
 
 // genfloatf sqrt (genfloatf x)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> sqrt(T x) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> sqrt(T x) __NOEXC {
   return __sycl_std::__invoke_native_sqrt<T>(x);
 }
 
 // genfloatf tan (genfloatf x)
 template <typename T>
-detail::enable_if_t<detail::is_genfloatf<T>::value, T> tan(T x) __NOEXC {
+detail::enable_if_t<detail::is_svgenfloatf<T>::value, T> tan(T x) __NOEXC {
   return __sycl_std::__invoke_native_tan<T>(x);
 }
 
 } // namespace native
 namespace half_precision {
 /* ----------------- 4.13.3 Math functions. ---------------------------------*/
+#define __SYCL_HALF_PRECISION_MATH_FUNCTION_OVERLOAD(NAME)                     \
+  template <size_t N>                                                          \
+  inline __SYCL_ALWAYS_INLINE marray<float, N> NAME(marray<float, N> x)        \
+      __NOEXC {                                                                \
+    marray<float, N> res;                                                      \
+    for (size_t i = 0; i < N / 2; i++) {                                       \
+      auto partial_res = __sycl_std::__invoke_half_##NAME<vec<float, 2>>(      \
```

Comment:
Since we know these are floats, is there a performance benefit to do the vector packing?

Reply:
Yes I think so: in theory we should be able to use vectorized loads/stores.

Reply:
I investigated this a bit further:

Using the OpenCL CPU backends vectorized loads are used in the final llvm IR when copying from marray to vector (including for float). It may be that the limitation of the loads/stores not being vectorized is only happening for NVPTX. I have therefore created an internal issue to try to fix this.

Some more information on the issue:

E.g. vector of size two implementation of tanh leads to this ptx for the fp16 case:

```
ld.global.b32 %hh1, [%rd4];
tanh.approx.f16x2 %hh2, %hh1;
mov.b32 {%h1, %h2}, %hh2;
st.global.v2.b16 [%rd4], {%h1, %h2};
```

Note this is little cleaner when using float since vec 2 loads for f32 are used rather than b32 for f16x2: then it will also use ld.global.v2.f32.

Whereas e.g. an marray of size two implementation of tanh (https://github.com/intel/llvm/pull/6038/files#diff-525a5c115806d68d90c778133cb3f137d979196033de778a085f30414e2bdf78R114) leads to (it should ideally match the above vector case):

 
```
ld.global.u16 %rd9, [%rd4];
ld.global.u16 %rd10, [%rd4+2];
shl.b64 %rd11, %rd10, 16;
or.b64 %rd12, %rd11, %rd9;
cvt.u32.u64 %r1, %rd12;
mov.b32 %hh1, %r1;
tanh.approx.f16x2 %hh3, %hh1;
mov.b32 {%h1, %h2}, %hh3;
st.global.b16 [%rd4], %h1;
st.global.b16 [%rd4+2], %h2;

```

From examining the llvm IR this appears to be because the compiler loses information about the copy from marray to vector during the optimization, then it loses the vectorized loads and stores. I think this is happening for all marray math functions.
I'm guessing that this is not something that is trivial to fix, and probably doesn't really make much difference. So I don't think this particular issue should block this PR.